lib/ansible/module_utils/k8s/raw.py

```diff
@@ -261,3 +296,56 @@ def create_project_request(self, definition):
         result['changed'] = True
         result['method'] = 'create'
         return result
+
+    def _wait_for(self, resource, name, namespace, predicate, timeout, state):
+        start = datetime.now()
+
+        def _wait_for_elapsed():
+            return (datetime.now() - start).seconds
+
+        response = None
+        while _wait_for_elapsed() < timeout:
+            try:
+                response = resource.get(name=name, namespace=namespace)
+                if predicate(response):
+                    return True, response.to_dict(), _wait_for_elapsed()
+                time.sleep(timeout // 20)
+            except NotFoundError:
+                if state == 'absent':
+                    return True, response.to_dict(), _wait_for_elapsed()
+        if response:
+            response = response.to_dict()
+        return False, response, _wait_for_elapsed()
+
+    def wait(self, resource, definition, timeout, state='present'):
+
+        def _deployment_ready(deployment):
+            # FIXME: frustratingly bool(deployment.status) is True even if status is empty
+            # Furthermore deployment.status.availableReplicas == deployment.status.replicas == None if status is empty
+            return (deployment.status and deployment.status.replicas is not None and
+                    deployment.status.availableReplicas == deployment.status.replicas and
+                    deployment.status.observedGeneration == deployment.metadata.generation)
+
+        def _pod_ready(pod):
+            return (pod.status and pod.status.containerStatuses is not None and
+                    all([container.ready for container in pod.status.containerStatuses]))
+
+        def _daemonset_ready(daemonset):
+            return (daemonset.status and daemonset.status.desiredNumberScheduled is not None and
+                    daemonset.status.numberReady == daemonset.status.desiredNumberScheduled and
+                    daemonset.status.observedGeneration == daemonset.metadata.generation)
+
+        def _resource_absent(resource):
+            return not resource
+
+        waiter = dict(
+            Deployment=_deployment_ready,
+            DaemonSet=_daemonset_ready,
+            Pod=_pod_ready
+        )
+        kind = definition['kind']
+        if state == 'present':
+            predicate = waiter.get(kind, lambda x: True)
```

Comment:
Is there any kind of generic predicate we could insert here? We could also allow the user to specify a predicate, as the `kubectl wait` command now does. We could also implement a wait function that attempts parity with `kubectl` in the client and just make use of it here, if you think that would make sense.

Reply:
Not that I could think of, everything behaves so differently. I'm not sure I want to overcomplicate this too much by allowing users to specify predicates, as then we'd likely need to go down the route of go specification handling.

The problem would still remain that the predicate only really works on the given resource kind (it would be nice to wait for endpoints to be up for a service to be ready)

Reply:
Yeah, the way kubectl did it was to allow the user to specify a condition in the status field of an object. I think (even if we don't add this now) we'll end up needing to allow some flexibility in the wait condition. We could also ignore kubectl wait and just allow the user to provide a JSON path and value, and we could just wait for the path to equal the value in the requested object. Regardless we should still have good defaults for the most common resources, so if we end up needing that flexibility we could probably just add another `wait_condition` object down the line.